### ISSUE TITLE
v0.5 support for 3.21.3 appears broken

### ISSUE THREAD

**Issue by bradleyhd** [hhvm/hhast#37]
I've been developing with 3.23.2 locally but we've finally rolled out 3.21.3 on the boxes where our linter is run. I've been trying to install both hhast and our repo that includes hhast as a dep, but get the following when trying to install:
```
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for facebook/definition-finder v1.6.4 -> satisfiable by facebook/definition-finder[v1.6.4].
    - facebook/definition-finder v1.6.4 requires hhvm ~3.23 -> your HHVM version does not satisfy that requirement.
  Problem 2
    - Installation request for facebook/hack-codegen v3.0.3 -> satisfiable by facebook/hack-codegen[v3.0.3].
    - facebook/hack-codegen v3.0.3 requires hhvm >=3.23.0 -> your HHVM version does not satisfy that requirement.
  Problem 3
    - Installation request for hhvm/hhvm-autoload v1.5.4 -> satisfiable by hhvm/hhvm-autoload[v1.5.4].
    - hhvm/hhvm-autoload v1.5.4 requires hhvm ^3.23 -> your HHVM version does not satisfy that requirement.
  Problem 4
    - Installation request for hhvm/type-assert v3.2.0 -> satisfiable by hhvm/type-assert[v3.2.0].
    - hhvm/type-assert v3.2.0 requires hhvm ^3.23.0 -> your HHVM version does not satisfy that requirement.
  Problem 5
    - Installation request for 91carriage/phpunit-hhi 5.7.2 -> satisfiable by 91carriage/phpunit-hhi[5.7.2].
    - 91carriage/phpunit-hhi 5.7.2 requires hhvm >=3.23.0 -> your HHVM version does not satisfy that requirement.
  Problem 6
    - hhvm/hhvm-autoload v1.5.4 requires hhvm ^3.23 -> your HHVM version does not satisfy that requirement.
    - hhvm/hsl v1.2.0 requires hhvm/hhvm-autoload ^1.4 -> satisfiable by hhvm/hhvm-autoload[v1.5.4].
    - Installation request for hhvm/hsl v1.2.0 -> satisfiable by hhvm/hsl[v1.2.0].
```
I know 3.21.3 probably isn't fun to support, so I've tried to fork and figure out a set of deps that does work up until v0.5 and got this far:
```
{
    "name": "hhvm/hhast",
    "bin": [ "bin/hhast-lint" ],
    "require-dev": {
        "facebook/hack-codegen": "=3.0.1",
        "phpunit/phpunit": "^5.7",
        "91carriage/phpunit-hhi": "=5.7.1",
        "facebook/fbexpect": "^0.3.0",
        "hhvm/hhvm-autoload": "=1.5.3",
        "facebook/definition-finder": "=1.6.3"
    },
    "require": {
        "hhvm": "^3.21.0",
        "hhvm/hsl": "=1.1.0",
        "hhvm/type-assert": "=3.1.0"
    }
}
```
but when I try to get hhast to lint itself:
```
bradley@slack:~/hhast$ hhvm bin/hhast-lint .

Warning: Destructor threw an object exception: exception 'HH\InvariantException' with message 'counter Facebook\HHAST\__Private\LinterCLI destroyed without calling ::end()' in /home/bradley/hhast/src/__Private/PerfCounter.php:51
Stack trace:
#0 /home/bradley/hhast/src/__Private/PerfCounter.php(51): HH\invariant_violation()
#1 /home/bradley/hhast/src/__Private/LinterCLI.php(155): Facebook\HHAST\__Private\PerfCounter->__destruct()
#2 /home/bradley/hhast/bin/hhast-lint(38): Facebook\HHAST\__Private\LinterCLI::mainAsync()
#3 {main} in /home/bradley/hhast/src/__Private/LinterCLI.php on line 155

Fatal error: Uncaught exception 'Exception' with message 'Invalid configuration file: /home/bradley/hhast/hhast-lint.json
  Expected type 'vec<T>', got type 'array'
Type trace:
#0 shape[roots]
' in /home/bradley/hhast/src/__Private/LinterCLIConfig.php:243
Stack trace:
#0 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(97): Facebook\HHAST\__Private\LinterCLIConfig::getConfigFromFile()
#1 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(98): Facebook\HHAST\__Private\LinterCLIConfig::getFromConfigFile$memoize_impl()
#2 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(121): Facebook\HHAST\__Private\LinterCLIConfig::getFromConfigFile()
#3 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(127): Facebook\HHAST\__Private\LinterCLIConfig::getForPathImpl$memoize_impl()
#4 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(109): Facebook\HHAST\__Private\LinterCLIConfig::getForPathImpl()
#5 /home/bradley/hhast/src/__Private/LinterCLI.php(138): Facebook\HHAST\__Private\LinterCLIConfig::getForPath()
#6 /home/bradley/hhast/bin/hhast-lint(38): Facebook\HHAST\__Private\LinterCLI->mainAsync()
#7 {main}
```
at this point I'm not sure what else to try. Any suggestions?

**Comment by fredemmott** [comment]
Can you double-check what version you have? That doesn't look like v0.5:

 - v0.5 had a clean test run, including self-lint, on 3.21: https://travis-ci.org/hhvm/hhast/jobs/326589829
 - the installed versions don't match your output
 - composer.lock at v0.5 does not specify those versions: https://github.com/hhvm/hhast/blob/v0.5/composer.lock
 - the change that sets the versions you mention ( https://github.com/hhvm/hhast/commit/bc5af15d82848a340c1788a40f1e60243ceef48f )  was created after v0.5

**Comment by bradleyhd** [comment]
```
bradley@slack:~/hhast$ hhvm composer.phar install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 33 installs, 0 updates, 0 removals
  - Installing fredemmott/hack-error-suppressor (v1.0.1): Loading from cache
  - Installing facebook/definition-finder (v1.6.3): Loading from cache
  - Installing hhvm/hhvm-autoload (v1.5.3): Loading from cache
  - Installing hhvm/hsl (v1.0.0): Loading from cache
  - Installing facebook/hack-codegen (v3.0.0): Loading from cache
  - Installing hhvm/type-assert (v3.1.0): Loading from cache
  - Installing symfony/yaml (v3.3.10): Loading from cache
  - Installing sebastian/version (2.0.1): Loading from cache
  - Installing sebastian/resource-operations (1.0.0): Loading from cache
  - Installing sebastian/recursion-context (2.0.0): Loading from cache
  - Installing sebastian/object-enumerator (2.0.1): Loading from cache
  - Installing sebastian/global-state (1.1.1): Loading from cache
  - Installing sebastian/exporter (2.0.0): Loading from cache
  - Installing sebastian/environment (2.0.0): Loading from cache
  - Installing sebastian/diff (1.4.3): Loading from cache
  - Installing sebastian/comparator (1.2.4): Loading from cache
  - Installing phpunit/php-text-template (1.2.1): Loading from cache
  - Installing doctrine/instantiator (1.0.5): Loading from cache
  - Installing phpunit/phpunit-mock-objects (3.4.4): Loading from cache
  - Installing phpunit/php-timer (1.0.9): Loading from cache
  - Installing phpunit/php-file-iterator (1.4.2): Loading from cache
  - Installing sebastian/code-unit-reverse-lookup (1.0.1): Loading from cache
  - Installing phpunit/php-token-stream (1.4.11): Loading from cache
  - Installing phpunit/php-code-coverage (4.0.8): Loading from cache
  - Installing webmozart/assert (1.2.0): Loading from cache
  - Installing phpdocumentor/reflection-common (1.0.1): Loading from cache
  - Installing phpdocumentor/type-resolver (0.3.0): Loading from cache
  - Installing phpdocumentor/reflection-docblock (3.2.2): Loading from cache
  - Installing phpspec/prophecy (v1.7.2): Loading from cache
  - Installing myclabs/deep-copy (1.7.0): Loading from cache
  - Installing phpunit/phpunit (5.7.23): Loading from cache
  - Installing 91carriage/phpunit-hhi (5.7.1): Cloning 6864897f02 from cache
  - Installing facebook/fbexpect (v0.3): Loading from cache
symfony/yaml suggests installing symfony/console (For validating YAML files using the lint command)
sebastian/global-state suggests installing ext-uopz (*)
phpunit/php-code-coverage suggests installing ext-xdebug (^2.5.1)
phpunit/phpunit suggests installing ext-xdebug (*)
phpunit/phpunit suggests installing phpunit/php-invoker (~1.1)
91carriage/phpunit-hhi suggests installing 91carriage/dbunit-hhi (HHI files for use with DbUnit)
Generating autoload files
bradley@slack:~/hhast$ hhvm bin/hhast-lint .

Warning: Destructor threw an object exception: exception 'HH\InvariantException' with message 'counter Facebook\HHAST\__Private\LinterCLI destroyed without calling ::end()' in /home/bradley/hhast/src/__Private/PerfCounter.php:51
Stack trace:
#0 /home/bradley/hhast/src/__Private/PerfCounter.php(51): HH\invariant_violation()
#1 /home/bradley/hhast/src/__Private/LinterCLI.php(155): Facebook\HHAST\__Private\PerfCounter->__destruct()
#2 /home/bradley/hhast/bin/hhast-lint(38): Facebook\HHAST\__Private\LinterCLI::mainAsync()
#3 {main} in /home/bradley/hhast/src/__Private/LinterCLI.php on line 155

Fatal error: Uncaught exception 'Exception' with message 'Invalid configuration file: /home/bradley/hhast/hhast-lint.json
  Expected type 'vec<T>', got type 'array'
Type trace:
#0 shape[roots]
' in /home/bradley/hhast/src/__Private/LinterCLIConfig.php:243
Stack trace:
#0 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(97): Facebook\HHAST\__Private\LinterCLIConfig::getConfigFromFile()
#1 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(98): Facebook\HHAST\__Private\LinterCLIConfig::getFromConfigFile$memoize_impl()
#2 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(121): Facebook\HHAST\__Private\LinterCLIConfig::getFromConfigFile()
#3 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(127): Facebook\HHAST\__Private\LinterCLIConfig::getForPathImpl$memoize_impl()
#4 /home/bradley/hhast/src/__Private/LinterCLIConfig.php(109): Facebook\HHAST\__Private\LinterCLIConfig::getForPathImpl()
#5 /home/bradley/hhast/src/__Private/LinterCLI.php(138): Facebook\HHAST\__Private\LinterCLIConfig::getForPath()
#6 /home/bradley/hhast/bin/hhast-lint(38): Facebook\HHAST\__Private\LinterCLI->mainAsync()
#7 {main}
```
agh yeah I must've done something wrong, thanks. but self-lint (or lint on our repo) still seem to have the vec/array type error

**Comment by bradleyhd** [comment]
Not sure if it matters but the HHVM version is a custom build
```
HipHop VM 3.21.3 (rel)
Compiler: remotes/origin/3.21.3-slack-0-blahblahblahblah
Repo schema: 90ed4216b98563d0585cf5c87959dc9ff7c2e1a1
```

**Comment by fredemmott** [comment]
- what's the JSON syntax for your roots? ideally full copy-paste of the file, or at least the 'roots' part
 - what does this get you in `hhvm -m debug`? `=json_decode('{"roots": ["foo", "bar"]}', true, 512, JSON_FB_HACK_ARRAYS)`

**Comment by fredemmott** [comment]
- it looks like an HHVM bug (in all versions) that we don't support JSON_FB_HACK_ARRAYS if built with USE_JSONC
 - if you're using a custom JSON parser library, you need to add support for JSON_FB_HACK_ARRAYS

**Comment by bradleyhd** [comment]
found it: `-DUSE_JSONC=ON` 😢 . this is really good stuff to know, thank you for all the help!

**Comment by fredemmott** [comment]
fwiw, if you're able to send a patch for facebook/hhvm#8098, I'll happily cherry pick it to 3.21 :)

**Comment by fredemmott** [comment]
https://github.com/facebook/hhvm/commit/94f3323bdfc219f66c0c52efa393612829704ec8

**Comment by fredemmott** [comment]
Fix is in 3.21.6: https://hhvm.com/blog/2018/01/30/hhvm-3.24.1.html

**Comment by bradleyhd** [comment]
🙇 can't think you enough, we're rolling this out now
